File: proxy-server.js
/**
 * TunnelVision Proxy Backend
 *
 * WHY WE NEED THIS:
 * Service Workers run in the browser. Even though they can intercept requests,
 * their own fetch() calls still originate from the browser's security context
 * (localhost:4600). Browsers enforce CORS on all cross-origin fetch calls —
 * including those from Service Workers — because the SW shares the page origin.
 *
 * A backend server's fetch() has NO such restriction. Node's http/fetch runs
 * outside the browser entirely, so it can fetch any URL without CORS headers.
 *
 * ARCHITECTURE:
 *
 *   Browser (page)
 *     └─ loads iframe at /student/TVP/proxy/<encoded-url>
 *          └─ Service Worker intercepts the request
 *               └─ SW calls: GET http://localhost:4601/fetch?url=<encoded-url>
 *                    └─ This server fetches the real URL (no CORS restrictions)
 *                         └─ Returns response with Access-Control-Allow-Origin: *
 *                              └─ SW receives body, rewrites HTML, returns to iframe
 *
 * Run with:  node proxy-server.js
 */

const http = require('http');
const https = require('https');
const { URL } = require('url');

const PORT = 4601;

// Headers we strip from the proxied response before returning to the browser.
// These would either block embedding or reveal internals.
const STRIP_RESPONSE_HEADERS = new Set([
  'content-security-policy',
  'content-security-policy-report-only',
  'x-frame-options',
  'x-content-type-options',
  'strict-transport-security',   // Don't forward HSTS to our local dev origin
  'transfer-encoding',           // Node handles this automatically
  'connection',
  'keep-alive',
]);

const server = http.createServer((req, res) => {
  // CORS headers so the Service Worker (different port = different origin) can
  // call this endpoint. Without these, the browser would block the SW's fetch.
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  // Handle CORS preflight
  if (req.method === 'OPTIONS') {
    res.writeHead(204);
    res.end();
    return;
  }

  const reqURL = new URL(req.url, `http://localhost:${PORT}`);

  if (reqURL.pathname !== '/fetch') {
    res.writeHead(404);
    res.end('TunnelVision proxy server. Use: GET /fetch?url=<encoded-url>');
    return;
  }

  const targetParam = reqURL.searchParams.get('url');
  if (!targetParam) {
    res.writeHead(400);
    res.end('Missing ?url= parameter');
    return;
  }

  let targetURL;
  try {
    targetURL = new URL(decodeURIComponent(targetParam));
  } catch {
    res.writeHead(400);
    res.end('Invalid URL');
    return;
  }

  // Only allow http and https — no file://, no internal network shenanigans
  if (targetURL.protocol !== 'http:' && targetURL.protocol !== 'https:') {
    res.writeHead(400);
    res.end('Only http/https URLs are supported');
    return;
  }

  console.log(`[TunnelVision] → ${targetURL.href}`);
  proxyRequest(targetURL, req, res);
});

/**
 * The actual proxying. This is what makes the backend essential:
 * Node's http/https.request has no CORS restrictions — it's just a TCP
 * connection to port 80 or 443, same as curl or any other HTTP client.
 */
function proxyRequest(targetURL, clientReq, clientRes) {
  const isHttps = targetURL.protocol === 'https:';
  const lib = isHttps ? https : http;

  const options = {
    hostname: targetURL.hostname,
    port: targetURL.port || (isHttps ? 443 : 80),
    path: targetURL.pathname + targetURL.search,
    method: clientReq.method || 'GET',
    headers: buildForwardHeaders(targetURL),
    // Follow redirects manually so we can rewrite the Location header
    // (auto-follow would lose our proxy wrapping on redirect targets)
    timeout: 10000,
  };

  const proxyReq = lib.request(options, (proxyRes) => {
    console.log(`[TunnelVision] ← ${proxyRes.statusCode} ${targetURL.href}`);

    // Handle redirects: rewrite Location so the browser stays in the proxy
    if (proxyRes.statusCode >= 300 && proxyRes.statusCode < 400 && proxyRes.headers.location) {
      const redirectTarget = new URL(proxyRes.headers.location, targetURL).href;
      clientRes.writeHead(302, {
        'Access-Control-Allow-Origin': '*',
        'Location': '/fetch?url=' + encodeURIComponent(redirectTarget),
        'Content-Type': 'text/plain',
      });
      clientRes.end();
      return;
    }

    // Build clean response headers
    const outHeaders = { 'Access-Control-Allow-Origin': '*' };
    for (const [key, value] of Object.entries(proxyRes.headers)) {
      if (!STRIP_RESPONSE_HEADERS.has(key.toLowerCase())) {
        outHeaders[key] = value;
      }
    }

    clientRes.writeHead(proxyRes.statusCode, outHeaders);

    // Pipe the response body directly — no buffering needed for binary content.
    // The Service Worker handles HTML/CSS rewriting on its end.
    proxyRes.pipe(clientRes);
  });

  proxyReq.on('error', (err) => {
    console.error(`[TunnelVision] Error fetching ${targetURL.href}: ${err.message}`);
    if (!clientRes.headersSent) {
      clientRes.writeHead(502, {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'text/plain',
      });
    }
    clientRes.end(`Proxy error: ${err.message}`);
  });

  proxyReq.on('timeout', () => {
    proxyReq.destroy();
    if (!clientRes.headersSent) {
      clientRes.writeHead(504, {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'text/plain',
      });
    }
    clientRes.end('Gateway timeout');
  });

  // Forward request body for POST/PUT
  clientReq.pipe(proxyReq);
}

/**
 * Build headers to send to the target site.
 * We masquerade as a real browser so sites respond normally.
 */
function buildForwardHeaders(targetURL) {
  return {
    'Host': targetURL.host,
    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
    'Accept-Language': 'en-US,en;q=0.9',
    'Accept-Encoding': 'gzip, deflate, br',
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    'Upgrade-Insecure-Requests': '1',
    // NOT forwarding: Cookie, Authorization — don't leak credentials to arbitrary sites
    // NOT forwarding: Referer — don't reveal we're proxying
  };
}

server.listen(PORT, () => {
  console.log(`
╔══════════════════════════════════════════╗
║   TunnelVision Proxy Backend             ║
║   http://localhost:${PORT}                  ║
╠══════════════════════════════════════════╣
║  GET /fetch?url=<encoded-url>            ║
║                                          ║
║  Keep this running alongside Jekyll.     ║
║  The Service Worker routes all requests  ║
║  through this server to bypass CORS.     ║
╚══════════════════════════════════════════╝
`);
});

File: proxy-server.py
"""
TunnelVision Proxy Backend (Flask)

WHY WE NEED THIS:
  Service Workers run inside the browser and share its origin (localhost:4600).
  The browser enforces CORS on all cross-origin fetch() calls, even from SWs.
  Python's requests library has no such restriction — it's a plain TCP client.

ARCHITECTURE:
  Browser → SW intercepts /TVP/proxy/<url>
           → SW calls GET http://localhost:4601/fetch?url=<url>
           → Flask fetches the real URL with requests (no CORS)
           → Returns response with Access-Control-Allow-Origin: *
           → SW rewrites HTML/CSS URLs → iframe displays the page

Run with:  python3 proxy-server.py
"""

from flask import Flask, request, Response
import requests
import urllib3
import time

urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

app = Flask(__name__)

# Simple in-memory cache to avoid hammering CDNs (causes 429s).
# Key: URL, Value: (timestamp, status, headers, body_bytes)
_cache = {}
CACHE_TTL = 60  # seconds — short enough to stay fresh, long enough to help

STRIP_HEADERS = {
    'content-security-policy',
    'content-security-policy-report-only',
    'x-frame-options',
    'x-content-type-options',
    'strict-transport-security',
    'transfer-encoding',
    'connection',
    # content-encoding MUST be stripped: requests decompresses the body but
    # leaves the header, so the browser would try to decompress raw bytes → ERR_FAILED.
    'content-encoding',
    # content-length changes after decompression — drop it or Flask recomputes it.
    'content-length',
}

@app.after_request
def add_cors(response):
    response.headers['Access-Control-Allow-Origin'] = '*'
    response.headers['Access-Control-Allow-Methods'] = 'GET, POST, OPTIONS'
    response.headers['Access-Control-Allow-Headers'] = 'Content-Type'
    return response

@app.route('/health')
def health():
    return {'status': 'ok', 'name': 'TunnelVision'}

@app.route('/fetch')
def proxy_fetch():
    target = request.args.get('url', '')
    if not target:
        return 'Missing ?url= parameter', 400

    if not target.startswith(('http://', 'https://')):
        return 'Only http/https URLs are supported', 400

    # Return cached response if still fresh.
    # This prevents 429 rate-limits when a page has many assets (images, CSS, fonts)
    # that all hit the same CDN host in a short burst.
    cached = _cache.get(target)
    if cached and (time.time() - cached['ts']) < CACHE_TTL:
        print(f'[TunnelVision] cache hit: {target}')
        return Response(cached['body'], status=cached['status'],
                        headers=cached['headers'],
                        content_type=cached['content_type'])

    print(f'[TunnelVision] → {target}')
    try:
        resp = requests.get(
            target,
            headers=build_headers(target),
            allow_redirects=True,
            timeout=10,
            verify=False,
            # Don't stream — we need to buffer to cache, and most assets are small.
            # For large video/binary files this is less ideal, but fine for web pages.
            stream=False,
        )
    except requests.exceptions.RequestException as e:
        return f'Proxy error: {e}', 502

    print(f'[TunnelVision] ← {resp.status_code} {target}')

    out_headers = {
        k: v for k, v in resp.headers.items()
        if k.lower() not in STRIP_HEADERS
    }
    content_type = resp.headers.get('Content-Type', 'application/octet-stream')

    # Cache successful responses (2xx). Don't cache errors.
    if 200 <= resp.status_code < 300:
        _cache[target] = {
            'ts': time.time(),
            'status': resp.status_code,
            'headers': out_headers,
            'content_type': content_type,
            'body': resp.content,
        }

    return Response(resp.content, status=resp.status_code,
                    headers=out_headers, content_type=content_type)

def build_headers(target_url):
    from urllib.parse import urlparse
    parsed = urlparse(target_url)
    return {
        'Host': parsed.netloc,
        'User-Agent': (
            'Mozilla/5.0 (Windows NT 10.0; Win64; x64) '
            'AppleWebKit/537.36 (KHTML, like Gecko) '
            'Chrome/120.0.0.0 Safari/537.36'
        ),
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.9',
        # Don't request gzip — let requests get plain text so there's no
        # content-encoding mismatch to worry about at all.
        'Accept-Encoding': 'identity',
        'Upgrade-Insecure-Requests': '1',
    }

if __name__ == '__main__':
    print("""
╔══════════════════════════════════════════╗
║   TunnelVision Proxy Backend (Flask)     ║
║   http://localhost:4601                  ║
╠══════════════════════════════════════════╣
║  GET /fetch?url=<encoded-url>            ║
║                                          ║
║  Run alongside Jekyll:                   ║
║    python3 proxy-server.py               ║
╚══════════════════════════════════════════╝
""")
    app.run(port=4601, debug=False)
